### PR TITLE
Task 9

### PR DESCRIPTION
**Pull request by vpuzan**
Task-9: Add parameters to the query

some clarification on the request parameters:
startDate - the beginning of the period for receiving meteorites (if absent, the first day of the week is expected)
endDate - the end of the period for receiving a meteor (if absent, the last day of the week is expected)
count - true or false, whether to count the total number of meteorites this week
wereDangerousMeteors - true or false, whether to search for dangerous meteorites or not, if true, it will display a list of these meteorites

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1,3 @@
+NASA_API_KEY=key
+NASA_API_URL=https://api.nasa.gov/neo/rest/v1/feed
+PORT=4000
```
Comment (mdziarkach) [inline .env]: please don't commit .env files to repository
this file should be in .gitignore
but you need to add .env.example with same structure as in .env file

File: delivery/server.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import express from 'express';
+import {getSortedAndFilteredMeteors} from '../use_case/meteors.js';
+import dotenv from "dotenv";
+
+dotenv.config({path: './resources/.env'});
```
Comment (mdziarkach) [inline delivery/server.js]: you will not have .env file on production and all keys will come from the environment 

```js
dotenv.config();
```
should be enough

File: delivery/server.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+const server = express();
+const PORT = process.env.PORT || 3000;
```
Comment (mdziarkach) [inline delivery/server.js]: please create config.js file for managing env variables

File: delivery/server.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+server.listen(PORT, () => {
```
Comment (mdziarkach) [inline delivery/server.js]: not a mistake but better to put it in the end of the file

File: delivery/server.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    console.log(`Server is running on port ${PORT}`);
+});
+
+server.get('/meteors', async (req, res) => {
+    try {
+        const {startDate, endDate, count, wereDangerousMeteors } = req.query;
+        const filteredData = await getSortedAndFilteredMeteors(startDate, endDate, count, wereDangerousMeteors);
```
Comment (mdziarkach) [inline delivery/server.js]: please convert all data inside of controller before providing them into the service

File: delivery/server.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        res.json(filteredData);
+    } catch (error) {
+        console.error('Error fetching data:', error);
+        res.status(500).json({error: 'Failed to fetch data'});
```
Comment (mdziarkach) [inline delivery/server.js]: next(error)

File: delivery/server.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    }
+});
```
Comment (mdziarkach) [inline delivery/server.js]: add `errorHandler` please somewhere there

here you can find a little bit more information https://medium.com/@mohsinogen/simplified-guide-to-setting-up-a-global-error-handler-in-express-js-daf8dd640b69

File: repository/nasa.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+const {NASA_API_URL, NASA_API_KEY} = process.env || (() => {
+    throw new Error('Missing env keys');
+})();
+
+const axiosConfig = {
+    httpsAgent: new https.Agent({
+        rejectUnauthorized: false, // switch off certificate for testing
+    }),
+};
```
Comment (mdziarkach) [inline repository/nasa.js]: move axios configuration into the separate file please 
add bese url

File: repository/nasa.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+export const fetchMeteors = async (startDate, endDate) => {
+    const apiUrl = `${NASA_API_URL}?start_date=${startDate}&end_date=${endDate}&api_key=${NASA_API_KEY}`;
+    const response = await axios.get(apiUrl, axiosConfig);
```
Comment (mdziarkach) [inline repository/nasa.js]: use can provide query params using axios params option

File: use_case/meteors.js [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    };
+}
+
+function getFirstDayOfWeek(date, firstDayOfWeek = 1) {
+    const day = date.getDay();
+    let diff = day - firstDayOfWeek;
+    if (diff < 0) {
+        diff -= 7;
+    }
+    return new Date(date.setDate(date.getDate() - diff));
```
Comment (mdziarkach) [inline use_case/meteors.js]: https://date-fns.org/v4.1.0/docs/startOfWeek

File: use_case/meteors.js [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+}
+function getLastDayOfWeek(date, lastDayOfWeek = 6) {
+    const day = date.getDay();
+    let diff = lastDayOfWeek - day;
+    if (diff <= 0) {
+        diff += 7;
+    }
+    return new Date(date.setDate(date.getDate() + diff));
```
Comment (mdziarkach) [inline use_case/meteors.js]: https://date-fns.org/v4.1.0/docs/endOfWeek

File: use_case/meteors.js [hunk 12]
```diff
@@ -0,0 +1,110 @@
+import {fetchMeteors} from '../repository/nasa.js';
+import {format} from "date-fns";
+
+export const getSortedAndFilteredMeteors = async (startDateQuery, endDateQuery, count, wereDangerousMeteors) => {
+    let {startDate, endDate} = setUpDatePeriod(startDateQuery, endDateQuery);
+    const response = await fetchMeteors(startDate, endDate);
+    let filteredMeteorData = mainInfoFilter(response);
+    pushMeteorsCount(count, filteredMeteorData, response);
+    pushDengerousMeteors(wereDangerousMeteors, filteredMeteorData);
```
Comment (mdziarkach) [inline use_case/meteors.js]: don't mutate data please
it is really hard to understand the code

File: resources/.env.example [hunk 13]
```diff
@@ -0,0 +1,3 @@
+NASA_API_KEY=your personal nasa api key
+NASA_API_URL=nasa api url
+PORT=port
```
Comment (mdziarkach) [inline resources/.env.example]: this file should be in the root directory

File: delivery/server.js [hunk 14]
```diff
@@ -0,0 +1,36 @@
+import express from 'express';
+import {getSortedAndFilteredMeteors} from '../use_case/meteors.js';
+import {config} from '../config/config.js';
+import Exception from "../exception/exception.js";
+import errorHandler from "../exception/error_middleware.js";
+
+const server = express();
+
+function isBoolean(str) {
```
Comment (mdziarkach) [inline delivery/server.js]: looks like this function should be somewhere in utils but not in the controller

File: delivery/server.js [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+server.get('/meteors', async (req, res, next) => {
+    try {
+        const {startDate, endDate, count, wereDangerousMeteors } = req.query;
+        const filteredData = await getSortedAndFilteredMeteors(startDate, endDate, isBoolean(count), isBoolean(wereDangerousMeteors));
+        res.json(filteredData);
+    } catch (error) {
+        next(new Exception(error.status, `Error fetching data: ${error.message}`));
+    }
+});
```
Comment (mdziarkach) [inline delivery/server.js]: let's split root server file and controller

File: utils/date_utils.js [hunk 16]
```diff
@@ -0,0 +1,20 @@
+import { format } from "date-fns";
+import { startOfWeek } from "date-fns";
+import { endOfWeek } from "date-fns";
```
Comment (mdziarkach) [inline utils/date_utils.js]: import { format , startOfWeek , endOfWeek } from "date-fns";

Comment (mdziarkach) [reply utils/date_utils.js]: this is not mistake but a bit shorter

File: config/config.js [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+dotenv.config()
+
+const {NASA_API_URL, NASA_API_KEY,PORT} = process.env || (() => {
+    throw new Error('Missing env keys');
+})();
+
+export const config = {
+    port: PORT,
+    NASA_API_KEY: NASA_API_KEY,
+    nasaUrl: `${NASA_API_URL}`
```
Comment (mdziarkach) [inline config/config.js]: you have NASA_API_KEY and nasaUrl - theay are the same. Could you remove one of the keys?

**Review by mdziarkach — APPROVED** [review]
Looks good
Only one not critical comment

File: config/config.js (hunk 17 above)
Comment (vpuzan) [reply config/config.js]: fixed

**Pull request merged**